### PR TITLE
Add unittests for updaters.

### PR DESCRIPTION
**Pull request by trivialfis**
This is to aid #3825.  Without unittests it's very hard for me to do refactoring since I can't interpret errors in r-tests.

### CODE REVIEW

**Comment by trivialfis** [comment]
Note to myself, parameter `int colmat_dtype` is removed.

**Comment by hcho3** [comment]
@trivialfis This is great! Let me know if you want me to review.

**Comment by trivialfis** [comment]
@hcho3 I'm afraid there is not much I can do.  It's very hard for me to come up with proper tests for these algorithms now.  
For simple one like `prune`, it's still possible to invent some data for specific testing cases, but for more complicate algorithms like `fast_hist`, it's best to come up with these test cases before or during development.
What I'm doing for `fast_hist` is just assuming it's correct, then create a simple test to "lock" the current output of it so that future changes won't be able to break it.

**Comment by hcho3** [comment]
@trivialfis I can add some tests for fast_hist after 0.81 release. Can we hold a bit?

**Comment by trivialfis** [comment]
@hcho3 That would be great. In the mean while can you help taking a look into these thing in this PR?
- Renamed fast_hist file into quantile_hist, spited the interface and impl,  removed fast_hist parameter.
- The current method of adding unittest for fast_hist. (By inheritance).
If you find these unsuitable, I can make changes early.

**Comment by trivialfis** [comment]
@hcho3 If the testing framework for `fast_hist` created by this PR is suitable, you can add your tests based on this PR,  otherwise you might duplicate the work.

**Comment by hcho3** [comment]
@trivialfis Okay, I will review this PR as it stands, so that we have a proper testing infrastructure in place. We can merge this PR after 0.81 release, which is set to happen on November 1, 2018.

**Comment by trivialfis** [comment]
@hcho3 Thanks!

File: src/logging.cc [hunk 1]
```diff
@@ -4,9 +4,9 @@
  * \brief Implementation of loggers.
  * \author Tianqi Chen
  */
+#include <rabit/rabit.h>
```
Comment (hcho3) [inline src/logging.cc]: So you are replacing `common/sync.h` with `rabit/rabit.h`?

Comment (trivialfis) [reply src/logging.cc]: This is all the code contained in original commom/sync.h:
```
#ifndef XGBOOST_COMMON_SYNC_H_
#define XGBOOST_COMMON_SYNC_H_
#include <rabit/rabit.h>
#endif  // XGBOOST_COMMON_SYNC_H_
```
Is there any reason to keep it?  If so I will revert this change and add a note to the code.

File: src/tree/param.h [hunk 2]
```diff
@@ -81,6 +81,23 @@ struct TrainParam : public dmlc::Parameter<TrainParam> {
   int gpu_batch_nrows;
   // the criteria to use for ranking splits
   std::string split_evaluator;
+
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply src/tree/param.h]: @hcho3 I have been going forth and back for this.  It seems all parameters from other updaters are grouped into this TrainParam except for `FastHist` which has an independent file containing parameters.  There is no practical reason for this change, just happened when re-organizing the code.

[the remark this answers is not in the document]
Comment (hcho3) [reply src/tree/param.h]: @trivialfis Got it. I don't particular lean on keep it separate either, so just leave it.

File: tests/cpp/tree/test_quantile_hist.cc [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+    void TestBuildHist(int nid,
+                       const GHistIndexMatrix& gmat,
+                       const DMatrix& fmat,
+                       const RegTree& tree) {
+      std::vector<GradientPair> gpair =
+          { {0.23f, 0.24f}, {0.24f, 0.25f}, {0.26f, 0.27f}, {0.27f, 0.28f},
+            {0.27f, 0.29f}, {0.37f, 0.39f}, {0.47f, 0.49f}, {0.57f, 0.59f} };
+      RealImpl::InitData(gmat, gpair, fmat, tree);
+      GHistIndexBlockMatrix quantile_index_block;
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply tests/cpp/tree/test_quantile_hist.cc]: I will leave it to anther PR dedicated to changing the code, maybe a later version of #3825.  This one is to make sure later PRs don't break anything. :) (Hopefully)

File: src/tree/updater_quantile_hist.cc [hunk 4]
```diff
@@ -0,0 +1,737 @@
...
+          }
+        }
+        best.Update(loss_chg, fid, split_pt, d_step == -1);
+      }
+    }
+  }
+  p_best->Update(best);
+}
+
+XGBOOST_REGISTER_TREE_UPDATER(FastHistMaker, "grow_fast_histmaker")
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply src/tree/updater_quantile_hist.cc]: @hcho3  I will rename the method and the class, add a `deprecated` warning to old name of this updater.  Does it seem reasonable to you?

File: tests/cpp/tree/test_quantile_hist.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (hcho3) [reply tests/cpp/tree/test_quantile_hist.cc]: Got it.

File: src/tree/updater_quantile_hist.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (hcho3) [reply src/tree/updater_quantile_hist.cc]: @trivialfis Yes. Deprecating `grow_fast_histmaker` is fine, since it was never a great name anyway.

File: src/logging.cc (hunk 1 above)
Comment (hcho3) [reply src/logging.cc]: Let's get rid of it then.

**Comment by hcho3** [comment]
I will now go ahead and start working on adding unit tests for 'hist'. Thanks!
